keps/407-gang-scheduling/README.md

```diff
@@ -0,0 +1,407 @@
+      - [Unit tests](#unit-tests)
+      - [Integration tests](#integration-tests)
+      - [e2e tests](#e2e-tests)
+  - [Graduation Criteria](#graduation-criteria)
+- [Implementation History](#implementation-history)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+## Summary
+
+<!--
+This section is incredibly important for producing high-quality, user-focused
+documentation such as release notes or a development roadmap. It should be
+possible to collect this information before implementation begins, in order to
+avoid requiring implementors to split their attention between writing release
+notes and implementing the feature itself. KEP editors and SIG Docs
+should help to ensure that the tone and content of the `Summary` section is
+useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+
+Both in this section and below, follow the guidelines of the [documentation
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+This KEP aims to enable gang scheduling support for LeaderWorkerSet by creating a PodGroup for each replica.
+This enhancement allows custom schedulers(e.g. Volcano, coscheduling scheduler-plugin, YuniKorn etc.) to perform 
+gang scheduling on the leader and its worker pods as a single unit, ensuring they are scheduled together to
+avoid deadlocks.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+Without gang scheduling support, partial scheduling may lead to resource deadlocks. 
+As reported in [issue #167](https://github.com/kubernetes-sigs/lws/issues/167), when cluster resources are limited, 
+the scheduler may prioritize scheduling leader pods while leaving their worker pods pending. 
+This results in a situation where no inference service can serve properly, as each service requires both leader and worker pods to be running.
+Gang scheduling support would ensure that pods within a replica are either all scheduled together or not scheduled at all, 
+thus preventing resource deadlocks and maintaining service availability.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+- Create different PodGroup resources for each lws replica for different custom schedulers (Volcano, scheduler-plugins, YuniKorn etc.) to enable gang scheduling
```

Comment:
I would again not scope this to one specific project, but to have that as an example. Other projects that add examples make the KEP stronger I think?

Reply:
Other projects require gang scheduling, such as spark, flink, training job mpi, etc., but other projects have already integrated gang scheduling. My KEP is only for the lws project, so I think the goal here is reasonable, right?

Reply:
I came to this KEP from an upstream (more general) link and I thought it was being referenced as a general solution (that happened to be under this repo). My misunderstanding. I would like to see a community maintained solution.

Reply:
This https://github.com/kubernetes/enhancements/issues/4671 can be the general solution.

Reply:
Yes, this is the general solution, but what about the custom scheduler? Can't https://github.com/kubernetes/enhancements/issues/4671 be the default implementation, while other scheduler can be configurable :)?